### PR TITLE
Migrate from tweens to clutter animations (fixes #92)

### LINKED ISSUE #92: Not working on 3.38.0
As gnome-shell has transitioned from using tweener to clutter animations and tweener was dropped, random wallpaper fails to load on 3.38.0 with the message `No JS module 'tweener' found in search path`

### PR DESCRIPTION
**Pull request by evfool**
Updated to work with the latest GNOME shell (3.38.0), and simplified the code using clutter animations.

### CODE REVIEW

**Review by ifl0w — CHANGES_REQUESTED** [review]
Hey, that seems to be a great PR already. Thanks for the help!
These transitions have been hacky since the beginning.
Please consider some tiny changes and I'll merge the PR.

Unfortunately, the animations only work correctly for the first request on my computer. However, I guess this is unrelated to your changes, as it was buggy for me previously already. 
Did you experience the same problem?

File: randomwallpaper@iflow.space/elements.js [hunk 1]
```diff
@@ -237,45 +236,23 @@ class StatusElement {
...
-				return true;
-			}
-		}
-
 	}
 
 	startLoading() {
-		this.isLoading = true;
-		Tweener.addTween(this.icon, this.loadingTweenOut);
+		this.isLoading = true
```
[the remark this answers is not in the document]
Comment (evfool) [reply randomwallpaper@iflow.space/elements.js]: isLoading field removed.

File: randomwallpaper@iflow.space/elements.js [hunk 2]
```diff
@@ -237,45 +236,23 @@ class StatusElement {
...
-			time: 2,
-			transition: 'easeInOutSine',
-			onComplete: function () {
-				try {
-					Tweener.addTween(_this.icon, _this.loadingTweenOut);
-				} catch (e) {
-					// swollow (not really important)
-				}
-			}
+			duration: 500,
```
[the remark this answers is not in the document]
Comment (evfool) [reply randomwallpaper@iflow.space/elements.js]: Previous animation durations were 1000 for fading in and 2000 for fading out, making up for a total of 3000ms for a cycle, now that we have auto-reverse the two parts are equal, so I have set 1500ms, resulting in a similar 3000ms fade-in+fade-out cycle duration. Not sure why it seemed ok when I tested with 500ms last time, and was too slow with larger values, now with 1500ms seems ok, and similar to the one I had before the changes.

**Comment by evfool** [comment]
Hi, thanks for the review, your changes requested are pushed to the branch. After pushing this PR I have realized that this way the extension looses backwards compatibility, and quickly checked, the thing introducing clutter animations (and recommending them to extension devs in gnome-shell HACKING.md) is gnome-shell!669, labeled with 3.34 release, meaning that if accepting this change, the extension will only work with versions 3.34+. Would you like to keep backwards compatibility (if there is such a thing) with older versions? Should I wire this code up only for 3.34+ and use the old code as a fallback on earlier versions?

**Comment by evfool** [comment]
Regarding the animation being stuck, I have just experienced it, and it looks like it's being caused by the transition (I had the extension enabled, deleted the extension folder, and invoked install.sh to install the development version) deleting the old wallpapers, and history not being able to find the old wallpaper to delete it. Should I handle the GIO error properly?
```
Oct 07 01:28:04 powerpc gnome-shell[1209]: JS ERROR: Gio.IOErrorEnum: Error removing file /home/evfool/.local/share/gnome-shell/extensions/randomwallpaper@iflow.space/wallpapers/1572507405968_wallhaven-w8jqo6.jpg: No such file or directory
                                           _deleteOldPictures@/home/evfool/.local/share/gnome-shell/extensions/randomwallpaper@iflow.space/history.js:151:15
                                           insert@/home/evfool/.local/share/gnome-shell/extensions/randomwallpaper@iflow.space/history.js:43:8
                                           fetchNewWallpaper/</</<@/home/evfool/.local/share/gnome-shell/extensions/randomwallpaper@iflow.space/wallpaperController.js:255:30
                                           _setBackground/</<@/home/evfool/.local/share/gnome-shell/extensions/randomwallpaper@iflow.space/wallpaperController.js:166:7
                                           _setPictureUriOfSettingsObject@/home/evfool/.local/share/gnome-shell/extensions/randomwallpaper@iflow.space/wallpaperController.js:199:6
                                           _setBackground/<@/home/evfool/.local/share/gnome-shell/extensions/randomwallpaper@iflow.space/wallpaperController.js:163:10
                                           _setPictureUriOfSettingsObject@/home/evfool/.local/share/gnome-shell/extensions/randomwallpaper@iflow.space/wallpaperController.js:199:6
                                           _setBackground@/home/evfool/.local/share/gnome-shell/extensions/randomwallpaper@iflow.space/wallpaperController.js:159:8
                                           fetchNewWallpaper/</<@/home/evfool/.local/share/gnome-shell/extensions/randomwallpaper@iflow.space/wallpaperController.js:253:10
                                           _fetchFile/<@/home/evfool/.local/share/gnome-shell/extensions/randomwallpaper@iflow.space/wallpaperController.js:144:5
```

**Comment by evfool** [comment]
@ifl0w, forgot to ask whether you could be experiencing the same issue, alas does clearing the history make the animations work (both start and stop) normally for subsequent wallpaper changes?

**Comment by ifl0w** [comment]
> Hi, thanks for the review, your changes requested are pushed to the branch. After pushing this PR I have realized that this way the extension looses backwards compatibility, and quickly checked, the thing introducing clutter animations (and recommending them to extension devs in gnome-shell HACKING.md) is https://gitlab.gnome.org/GNOME/gnome-shell/-/merge_requests/669, labeled with 3.34 release, meaning that if accepting this change, the extension will only work with versions 3.34+. Would you like to keep backwards compatibility (if there is such a thing) with older versions? Should I wire this code up only for 3.34+ and use the old code as a fallback on earlier versions?

The most recent version does support 3.34+ only, so this should be fine. Remaining backward compatible is quite some work as the gnome-shell APIs introduces breaking changes regularly. Older unsupported versions simply won't get new features. Important is that the versions in the `metadata.json` are really supported.

The file removal error is not related to the icon animation and should be handled separately, your changes in `history.js` are good and could be part of a separate PR, if you wish, but the error handling should be done in the `clear()` function as well.

Clearing the history and having invalid images in the history did not seem to influence the animation bug on my machine. After the first successful animation, consecutive animations consist of a single fade out until it is stopped and the opacity pops back to normal.
I'd assume that the animation error is due to some animation state not beeing reset correctly. Maybe there has to be some `save_easing_state`/`restore_easing_state`, or the `remove_all_transitions` call is simply not enough or does not work correcltly. I would have to look more into the Clutter animation system to find the problem. Of course, any hint is welcome and I do not have a lot of time to invest right now.

**Review by ifl0w — APPROVED** [review]
Looks great! Thanks for the work. I'll merge it and the changes will be in the next release in a couple of days/weeks.

**Comment by evfool** [comment]
> The most recent version does support 3.34+ only, so this should be fine. Remaining backward compatible is quite some work as the gnome-shell APIs introduces breaking changes regularly. Older unsupported versions simply won't get new features. Important is that the versions in the `metadata.json` are really supported.
OK.
 
> The file removal error is not related to the icon animation and should be handled separately, your changes in `history.js` are good and could be part of a separate PR, if you wish, but the error handling should be done in the `clear()` function as well.
> 
Maybe will try to do so.

> Clearing the history and having invalid images in the history did not seem to influence the animation bug on my machine. After the first successful animation, consecutive animations consist of a single fade out until it is stopped and the opacity pops back to normal.
> I'd assume that the animation error is due to some animation state not beeing reset correctly. Maybe there has to be some `save_easing_state`/`restore_easing_state`, or the `remove_all_transitions` call is simply not enough or does not work correcltly. I would have to look more into the Clutter animation system to find the problem. Of course, any hint is welcome and I do not have a lot of time to invest right now.

I will try to check why it works on my machines (both Gnome-Shell 3.36 and 3.38), animation pulses until it finishes. save_easing_state / restore_easing_state is handled internally by the .ease call, according to the _easeActor methor from https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/master/js/ui/environment.js, which is being called by .ease.

**Comment by ifl0w** [comment]
Are both of your setups running on Wayland? I'm currently not and it could make a difference as well. Need to test this later.

**Comment by evfool** [comment]
> You are right, could be a more difficult bug as well.
> Are both of your setups running on Wayland? I'm currently not and it could make a difference as well. Need to test this later.
3.38 is currently on Wayland (although when I made the fix was on X for easy shell reload), 3.36 is on Xorg.

**Pull request merged**